server/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -648,8 +650,27 @@ private IndexShard getShard(ShardSearchRequest request) {
         return indicesService.indexServiceSafe(request.shardId().getIndex()).getShard(request.shardId().id());
     }
 
-    private static <T> void runAsync(Executor executor, CheckedSupplier<T, Exception> executable, ActionListener<T> listener) {
-        executor.execute(ActionRunnable.supply(listener, executable));
```

Comment:
This was the problem, we implicitly assumed that the listener would always count down the reference here. This held for all current implementations which saw a `ChannelActionListener` but broke once I tried to use this with the multi-search response which can either go to a `ChannelActionListener` or some other listeners (which would then have to be aware of the fact that they need to count-down by one which would make the code needlessly brittle.